File: app/src/main/java/com/tools/vspace/engine/VEnvironment.java
package com.tools.vspace.engine;

import android.content.Context;
import android.util.Log;

import java.io.File;
import java.util.HashMap;
import java.util.Map;

/**
 * Virtual Environment manager for the app-level engine.
 * Wraps the blackbox-core VEnvironment and provides
 * additional virtual filesystem and property management.
 */
public class VEnvironment {

    private static final String TAG = "VEnvWrapper";

    private final Context context;
    private final Map<String, String> virtualProps = new HashMap<>();

    public VEnvironment(Context context) {
        this.context = context;
        initVirtualProps();
    }

    private void initVirtualProps() {
        virtualProps.put("ro.product.model", "Pixel 5");
        virtualProps.put("ro.product.manufacturer", "Google");
        virtualProps.put("ro.product.brand", "google");
        virtualProps.put("ro.product.device", "redfin");
        virtualProps.put("ro.product.name", "redfin");
        virtualProps.put("ro.build.display.id", "RQ3A.211001.001");
        virtualProps.put("ro.build.version.release", "11");
        virtualProps.put("ro.build.version.sdk", "30");
        virtualProps.put("ro.build.fingerprint",
                "google/redfin/redfin:11/RQ3A.211001.001/7641976:userdebug/dev-keys");
        virtualProps.put("ro.hardware", "redfin");
        virtualProps.put("ro.build.type", "userdebug");
        virtualProps.put("ro.build.tags", "dev-keys");
        virtualProps.put("ro.debuggable", "0");
        virtualProps.put("ro.secure", "1");
        Log.i(TAG, "Virtual properties initialized for Pixel 5");
    }

    public String getProperty(String key, String defaultValue) {
        String value = virtualProps.get(key);
        return value != null ? value : defaultValue;
    }

    public void setProperty(String key, String value) {
        virtualProps.put(key, value);
    }

    public File getVirtualDataDir(String packageName) {
        File dataDir = new File(context.getFilesDir(), "virtual_data/" + packageName);
        dataDir.mkdirs();
        return dataDir;
    }

    public File getVirtualBinDir() {
        File binDir = new File(context.getFilesDir(), "virtual_data/system/bin");
        binDir.mkdirs();
        return binDir;
    }

    public boolean shouldRedirect(String path) {
        if (path == null) return false;
        return path.startsWith("/system/bin/su") ||
               path.startsWith("/system/xbin/su") ||
               path.startsWith("/sbin/su");
    }

    public String getRedirectedPath(String originalPath) {
        if (originalPath == null) return null;
        if (originalPath.contains("/su")) {
            return new File(getVirtualBinDir(), "su").getAbsolutePath();
        }
        return originalPath;
    }
}

File: blackbox-core/src/main/java/com/tools/vspace/blackbox/VEnvironment.kt
package com.tools.vspace.blackbox

import android.content.Context
import android.content.SharedPreferences
import android.os.Build
import android.util.Log
import java.io.File

/**
 * Virtual Environment configuration and management
 * Handles virtual filesystem paths, device spoofing, and environment variables
 */
class VEnvironment(private val context: Context) {

    companion object {
        private const val TAG = "VEnvironment"
        private const val PREFS_NAME = "venvironment_config"

        // Default spoofed device: Google Pixel 5
        const val DEFAULT_MODEL = "Pixel 5"
        const val DEFAULT_MANUFACTURER = "Google"
        const val DEFAULT_BRAND = "google"
        const val DEFAULT_DEVICE = "redfin"
        const val DEFAULT_PRODUCT = "redfin"
        const val DEFAULT_HARDWARE = "redfin"
        const val DEFAULT_FINGERPRINT = "google/redfin/redfin:11/RQ3A.211001.001/7641976:userdebug/dev-keys"
        const val DEFAULT_BUILD_ID = "RQ3A.211001.001"
    }

    data class DeviceSpoof(
        val model: String = DEFAULT_MODEL,
        val manufacturer: String = DEFAULT_MANUFACTURER,
        val brand: String = DEFAULT_BRAND,
        val device: String = DEFAULT_DEVICE,
        val product: String = DEFAULT_PRODUCT,
        val hardware: String = DEFAULT_HARDWARE,
        val fingerprint: String = DEFAULT_FINGERPRINT,
        val buildId: String = DEFAULT_BUILD_ID
    )

    private val prefs: SharedPreferences = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    val virtualDataDir: File
        get() = File(context.filesDir, "virtual_data")

    val virtualSystemDir: File
        get() = File(virtualDataDir, "system")

    val virtualBinDir: File
        get() = File(virtualSystemDir, "bin")

    /**
     * Initialize virtual environment directories
     */
    fun initialize() {
        virtualDataDir.mkdirs()
        virtualSystemDir.mkdirs()
        virtualBinDir.mkdirs()

        // Create fake su binary path
        val suFile = File(virtualBinDir, "su")
        if (!suFile.exists()) {
            suFile.createNewFile()
            suFile.setExecutable(true)
        }

        Log.i(TAG, "Virtual environment initialized at: ${virtualDataDir.absolutePath}")
    }

    /**
     * Get current device spoof configuration
     */
    fun getDeviceSpoof(): DeviceSpoof {
        return DeviceSpoof(
            model = prefs.getString("spoof_model", DEFAULT_MODEL) ?: DEFAULT_MODEL,
            manufacturer = prefs.getString("spoof_manufacturer", DEFAULT_MANUFACTURER) ?: DEFAULT_MANUFACTURER,
            brand = prefs.getString("spoof_brand", DEFAULT_BRAND) ?: DEFAULT_BRAND,
            device = prefs.getString("spoof_device", DEFAULT_DEVICE) ?: DEFAULT_DEVICE,
            product = prefs.getString("spoof_product", DEFAULT_PRODUCT) ?: DEFAULT_PRODUCT,
            hardware = prefs.getString("spoof_hardware", DEFAULT_HARDWARE) ?: DEFAULT_HARDWARE,
            fingerprint = prefs.getString("spoof_fingerprint", DEFAULT_FINGERPRINT) ?: DEFAULT_FINGERPRINT,
            buildId = prefs.getString("spoof_build_id", DEFAULT_BUILD_ID) ?: DEFAULT_BUILD_ID
        )
    }

    /**
     * Save device spoof configuration
     */
    fun saveDeviceSpoof(spoof: DeviceSpoof) {
        prefs.edit().apply {
            putString("spoof_model", spoof.model)
            putString("spoof_manufacturer", spoof.manufacturer)
            putString("spoof_brand", spoof.brand)
            putString("spoof_device", spoof.device)
            putString("spoof_product", spoof.product)
            putString("spoof_hardware", spoof.hardware)
            putString("spoof_fingerprint", spoof.fingerprint)
            putString("spoof_build_id", spoof.buildId)
            apply()
        }
    }

    /**
     * Check if virtual environment is properly set up
     */
    fun isReady(): Boolean {
        return virtualDataDir.exists() && virtualBinDir.exists()
    }
}
